<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>References | Supplemental Material for Directed Digital Evolution Project</title>
  <meta name="description" content="Supplemental material" />
  <meta name="generator" content="bookdown 0.27 and GitBook 2.6.7" />

  <meta property="og:title" content="References | Supplemental Material for Directed Digital Evolution Project" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Supplemental material" />
  <meta name="github-repo" content="amlalejini/directed-digital-evolution" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="References | Supplemental Material for Directed Digital Evolution Project" />
  
  <meta name="twitter:description" content="Supplemental material" />
  

<meta name="author" content="Alexander Lalejini, Emily Dolson, Anya E. Vostinar, and Luis Zaman" />


<meta name="date" content="2022-06-23" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="varied-genome-lengths.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Supplemental material</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#about-this-supplemental-material"><i class="fa fa-check"></i><b>1.1</b> About this supplemental material</a><ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#contents"><i class="fa fa-check"></i><b>1.1.1</b> Contents</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#contributing-authors"><i class="fa fa-check"></i><b>1.2</b> Contributing authors</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#research-overview"><i class="fa fa-check"></i><b>1.3</b> Research overview</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="software-availability.html"><a href="software-availability.html"><i class="fa fa-check"></i><b>2</b> Software availability</a></li>
<li class="chapter" data-level="3" data-path="data-availability.html"><a href="data-availability.html"><i class="fa fa-check"></i><b>3</b> Data availability</a></li>
<li class="chapter" data-level="4" data-path="compiling-and-running-our-experiments.html"><a href="compiling-and-running-our-experiments.html"><i class="fa fa-check"></i><b>4</b> Compiling and running our experiments</a><ul>
<li class="chapter" data-level="4.1" data-path="compiling-and-running-our-experiments.html"><a href="compiling-and-running-our-experiments.html#manual"><i class="fa fa-check"></i><b>4.1</b> Manual</a></li>
<li class="chapter" data-level="4.2" data-path="compiling-and-running-our-experiments.html"><a href="compiling-and-running-our-experiments.html#docker"><i class="fa fa-check"></i><b>4.2</b> Docker</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="digital-organisms.html"><a href="digital-organisms.html"><i class="fa fa-check"></i><b>5</b> Digital Organisms</a><ul>
<li class="chapter" data-level="5.1" data-path="digital-organisms.html"><a href="digital-organisms.html#virtual-hardware-components"><i class="fa fa-check"></i><b>5.1</b> Virtual Hardware Components</a></li>
<li class="chapter" data-level="5.2" data-path="digital-organisms.html"><a href="digital-organisms.html#instruction-set"><i class="fa fa-check"></i><b>5.2</b> Instruction set</a></li>
<li class="chapter" data-level="5.3" data-path="digital-organisms.html"><a href="digital-organisms.html#ancestral-genomes"><i class="fa fa-check"></i><b>5.3</b> Ancestral genomes</a><ul>
<li class="chapter" data-level="5.3.1" data-path="digital-organisms.html"><a href="digital-organisms.html#ancestral-genome-for-genetic-programming-experiment"><i class="fa fa-check"></i><b>5.3.1</b> Ancestral genome for genetic programming experiment</a></li>
<li class="chapter" data-level="5.3.2" data-path="digital-organisms.html"><a href="digital-organisms.html#ancestral-genome-for-directed-evolution-experiments"><i class="fa fa-check"></i><b>5.3.2</b> Ancestral genome for directed evolution experiments</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html"><i class="fa fa-check"></i><b>6</b> Conventional genetic programming experiment</a><ul>
<li class="chapter" data-level="6.1" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#overview"><i class="fa fa-check"></i><b>6.1</b> Overview</a></li>
<li class="chapter" data-level="6.2" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#analysis-dependencies"><i class="fa fa-check"></i><b>6.2</b> Analysis dependencies</a></li>
<li class="chapter" data-level="6.3" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#setup"><i class="fa fa-check"></i><b>6.3</b> Setup</a></li>
<li class="chapter" data-level="6.4" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#number-of-successful-replicates"><i class="fa fa-check"></i><b>6.4</b> Number of successful replicates</a></li>
<li class="chapter" data-level="6.5" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#final-task-coverage"><i class="fa fa-check"></i><b>6.5</b> Final task coverage</a></li>
<li class="chapter" data-level="6.6" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#generation-2000-task-coverage"><i class="fa fa-check"></i><b>6.6</b> Generation 2,000 task coverage</a></li>
<li class="chapter" data-level="6.7" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#population-level-task-coverage"><i class="fa fa-check"></i><b>6.7</b> Population-level task coverage</a></li>
<li class="chapter" data-level="6.8" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#generations-elapsed-before-a-solution-evolves"><i class="fa fa-check"></i><b>6.8</b> Generations elapsed before a solution evolves</a></li>
<li class="chapter" data-level="6.9" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#best-task-coverage-over-time"><i class="fa fa-check"></i><b>6.9</b> Best task coverage over time</a></li>
<li class="chapter" data-level="6.10" data-path="conventional-genetic-programming-experiment.html"><a href="conventional-genetic-programming-experiment.html#manuscript-figure"><i class="fa fa-check"></i><b>6.10</b> Manuscript Figure</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html"><i class="fa fa-check"></i><b>7</b> Directed digital evolution experiment</a><ul>
<li class="chapter" data-level="7.1" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#overview-1"><i class="fa fa-check"></i><b>7.1</b> Overview</a></li>
<li class="chapter" data-level="7.2" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#analysis-dependencies-1"><i class="fa fa-check"></i><b>7.2</b> Analysis dependencies</a></li>
<li class="chapter" data-level="7.3" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#setup-1"><i class="fa fa-check"></i><b>7.3</b> Setup</a></li>
<li class="chapter" data-level="7.4" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#best-single-population-task-coverage"><i class="fa fa-check"></i><b>7.4</b> Best single-population task coverage</a><ul>
<li class="chapter" data-level="7.4.1" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#best-single-population-task-coverage-time-series"><i class="fa fa-check"></i><b>7.4.1</b> Best single-population task coverage time series</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#metapopulation-task-coverage"><i class="fa fa-check"></i><b>7.5</b> Metapopulation task coverage</a><ul>
<li class="chapter" data-level="7.5.1" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#metapopulation-task-coverage-time-series"><i class="fa fa-check"></i><b>7.5.1</b> Metapopulation task coverage time series</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#metapopulation-task-profile-diversity"><i class="fa fa-check"></i><b>7.6</b> Metapopulation task profile diversity</a><ul>
<li class="chapter" data-level="7.6.1" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#number-of-different-task-profiles"><i class="fa fa-check"></i><b>7.6.1</b> Number of different task profiles</a></li>
<li class="chapter" data-level="7.6.2" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#task-profile-spread"><i class="fa fa-check"></i><b>7.6.2</b> Task profile spread</a></li>
<li class="chapter" data-level="7.6.3" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#task-profile-entropy"><i class="fa fa-check"></i><b>7.6.3</b> Task profile entropy</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#task-coverage-per-n-populations"><i class="fa fa-check"></i><b>7.7</b> Task coverage per N populations</a></li>
<li class="chapter" data-level="7.8" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#average-number-of-different-populations-selected-per-generation"><i class="fa fa-check"></i><b>7.8</b> Average number of different populations selected per generation</a><ul>
<li class="chapter" data-level="7.8.1" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#entropy-of-selected-population-ids"><i class="fa fa-check"></i><b>7.8.1</b> Entropy of selected population IDs</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#average-number-of-organisms-in-populations-at-end-of-maturation-period"><i class="fa fa-check"></i><b>7.9</b> Average number of organisms in populations at end of maturation period</a></li>
<li class="chapter" data-level="7.10" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#average-generations-per-maturation-period"><i class="fa fa-check"></i><b>7.10</b> Average generations per maturation period</a></li>
<li class="chapter" data-level="7.11" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#representative-task-profiles"><i class="fa fa-check"></i><b>7.11</b> Representative task profiles</a></li>
<li class="chapter" data-level="7.12" data-path="directed-digital-evolution-experiment.html"><a href="directed-digital-evolution-experiment.html#manuscript-figures"><i class="fa fa-check"></i><b>7.12</b> Manuscript figures</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><i class="fa fa-check"></i><b>8</b> Aligned individual-level and population-level task directed evolution experiment</a><ul>
<li class="chapter" data-level="8.1" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#overview-2"><i class="fa fa-check"></i><b>8.1</b> Overview</a></li>
<li class="chapter" data-level="8.2" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#analysis-dependencies-2"><i class="fa fa-check"></i><b>8.2</b> Analysis dependencies</a></li>
<li class="chapter" data-level="8.3" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#setup-2"><i class="fa fa-check"></i><b>8.3</b> Setup</a></li>
<li class="chapter" data-level="8.4" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#best-single-population-task-coverage-1"><i class="fa fa-check"></i><b>8.4</b> Best single-population task coverage</a><ul>
<li class="chapter" data-level="8.4.1" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#best-single-population-task-coverage-time-series-1"><i class="fa fa-check"></i><b>8.4.1</b> Best single-population task coverage time series</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#metapopulation-task-coverage-1"><i class="fa fa-check"></i><b>8.5</b> Metapopulation task coverage</a><ul>
<li class="chapter" data-level="8.5.1" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#metapopulation-task-coverage-time-series-1"><i class="fa fa-check"></i><b>8.5.1</b> Metapopulation task coverage time series</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#metapopulation-task-profile-diversity-1"><i class="fa fa-check"></i><b>8.6</b> Metapopulation task profile diversity</a><ul>
<li class="chapter" data-level="8.6.1" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#number-of-different-task-profiles-1"><i class="fa fa-check"></i><b>8.6.1</b> Number of different task profiles</a></li>
<li class="chapter" data-level="8.6.2" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#task-profile-spread-1"><i class="fa fa-check"></i><b>8.6.2</b> Task profile spread</a></li>
<li class="chapter" data-level="8.6.3" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#task-profile-entropy-1"><i class="fa fa-check"></i><b>8.6.3</b> Task profile entropy</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#task-coverage-per-n-populations-1"><i class="fa fa-check"></i><b>8.7</b> Task coverage per N populations</a></li>
<li class="chapter" data-level="8.8" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#average-number-of-different-populations-selected-per-generation-1"><i class="fa fa-check"></i><b>8.8</b> Average number of different populations selected per generation</a><ul>
<li class="chapter" data-level="8.8.1" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#entropy-of-selected-population-ids-1"><i class="fa fa-check"></i><b>8.8.1</b> Entropy of selected population IDs</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#average-number-of-organisms-in-populations-at-end-of-maturation-period-1"><i class="fa fa-check"></i><b>8.9</b> Average number of organisms in populations at end of maturation period</a></li>
<li class="chapter" data-level="8.10" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#average-generations-per-maturation-period-1"><i class="fa fa-check"></i><b>8.10</b> Average generations per maturation period</a></li>
<li class="chapter" data-level="8.11" data-path="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html"><a href="aligned-individual-level-and-population-level-task-directed-evolution-experiment.html#manuscript-figures-1"><i class="fa fa-check"></i><b>8.11</b> Manuscript figures</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html"><i class="fa fa-check"></i><b>9</b> Varying population maturation period</a><ul>
<li class="chapter" data-level="9.1" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#overview-3"><i class="fa fa-check"></i><b>9.1</b> Overview</a></li>
<li class="chapter" data-level="9.2" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#analysis-dependencies-3"><i class="fa fa-check"></i><b>9.2</b> Analysis dependencies</a></li>
<li class="chapter" data-level="9.3" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#setup-3"><i class="fa fa-check"></i><b>9.3</b> Setup</a></li>
<li class="chapter" data-level="9.4" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#average-number-of-organisms"><i class="fa fa-check"></i><b>9.4</b> Average number of organisms</a></li>
<li class="chapter" data-level="9.5" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#average-generations-elapsed-during-the-maturation-period"><i class="fa fa-check"></i><b>9.5</b> Average generations elapsed during the maturation period</a></li>
<li class="chapter" data-level="9.6" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#total-generations"><i class="fa fa-check"></i><b>9.6</b> Total generations</a></li>
<li class="chapter" data-level="9.7" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#performance"><i class="fa fa-check"></i><b>9.7</b> Performance</a><ul>
<li class="chapter" data-level="9.7.1" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#cpu-cycles-per-replication"><i class="fa fa-check"></i><b>9.7.1</b> CPU cycles per replication</a></li>
<li class="chapter" data-level="9.7.2" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#best-population-task-coverage"><i class="fa fa-check"></i><b>9.7.2</b> Best-population task coverage</a></li>
<li class="chapter" data-level="9.7.3" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#metapopulation-task-coverage-2"><i class="fa fa-check"></i><b>9.7.3</b> Metapopulation task coverage</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#population-level-task-profile-diversity"><i class="fa fa-check"></i><b>9.8</b> Population-level Task Profile Diversity</a><ul>
<li class="chapter" data-level="9.8.1" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#task-profile-richness"><i class="fa fa-check"></i><b>9.8.1</b> Task profile richness</a></li>
<li class="chapter" data-level="9.8.2" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#task-profile-entropy-2"><i class="fa fa-check"></i><b>9.8.2</b> Task profile entropy</a></li>
<li class="chapter" data-level="9.8.3" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#spread-average-cosine-distance"><i class="fa fa-check"></i><b>9.8.3</b> Spread (average cosine distance)</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#selection"><i class="fa fa-check"></i><b>9.9</b> Selection</a><ul>
<li class="chapter" data-level="9.9.1" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#average-number-of-unique-populations-selected"><i class="fa fa-check"></i><b>9.9.1</b> Average number of unique populations selected</a></li>
<li class="chapter" data-level="9.9.2" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#average-entropy-of-selection-ids"><i class="fa fa-check"></i><b>9.9.2</b> Average entropy of selection ids</a></li>
</ul></li>
<li class="chapter" data-level="9.10" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#manuscript-figures-2"><i class="fa fa-check"></i><b>9.10</b> Manuscript Figures</a></li>
<li class="chapter" data-level="9.11" data-path="varying-population-maturation-period.html"><a href="varying-population-maturation-period.html#discussion"><i class="fa fa-check"></i><b>9.11</b> Discussion</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html"><i class="fa fa-check"></i><b>10</b> Varied genome lengths</a><ul>
<li class="chapter" data-level="10.1" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#overview-4"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#analysis-dependencies-4"><i class="fa fa-check"></i><b>10.2</b> Analysis dependencies</a></li>
<li class="chapter" data-level="10.3" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#setup-4"><i class="fa fa-check"></i><b>10.3</b> Setup</a></li>
<li class="chapter" data-level="10.4" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#average-generations-per-maturation-period-2"><i class="fa fa-check"></i><b>10.4</b> Average generations per maturation period</a></li>
<li class="chapter" data-level="10.5" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#total-generations-over-experiment"><i class="fa fa-check"></i><b>10.5</b> Total generations over experiment</a></li>
<li class="chapter" data-level="10.6" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#performance-1"><i class="fa fa-check"></i><b>10.6</b> Performance</a><ul>
<li class="chapter" data-level="10.6.1" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#cpu-cycles-per-replication-1"><i class="fa fa-check"></i><b>10.6.1</b> CPU cycles per replication</a></li>
<li class="chapter" data-level="10.6.2" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#best-single-population-task-coverage-2"><i class="fa fa-check"></i><b>10.6.2</b> Best single-population task coverage</a></li>
<li class="chapter" data-level="10.6.3" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#metapopulation-task-coverage-3"><i class="fa fa-check"></i><b>10.6.3</b> Metapopulation task coverage</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#population-level-task-profile-diversity-1"><i class="fa fa-check"></i><b>10.7</b> Population-level task profile diversity</a><ul>
<li class="chapter" data-level="10.7.1" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#task-profile-richness-1"><i class="fa fa-check"></i><b>10.7.1</b> Task profile richness</a></li>
<li class="chapter" data-level="10.7.2" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#task-profile-entropy-3"><i class="fa fa-check"></i><b>10.7.2</b> Task profile entropy</a></li>
<li class="chapter" data-level="10.7.3" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#spread-avg-cosine-distance"><i class="fa fa-check"></i><b>10.7.3</b> Spread (avg cosine distance)</a></li>
</ul></li>
<li class="chapter" data-level="10.8" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#selection-1"><i class="fa fa-check"></i><b>10.8</b> Selection</a><ul>
<li class="chapter" data-level="10.8.1" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#average-number-of-unique-populations-selected-1"><i class="fa fa-check"></i><b>10.8.1</b> Average number of unique populations selected</a></li>
<li class="chapter" data-level="10.8.2" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#average-entropy-of-selection-ids-1"><i class="fa fa-check"></i><b>10.8.2</b> Average entropy of selection ids</a></li>
</ul></li>
<li class="chapter" data-level="10.9" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#time-series"><i class="fa fa-check"></i><b>10.9</b> Time series</a><ul>
<li class="chapter" data-level="10.9.1" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#best-single-population-task-coverage-3"><i class="fa fa-check"></i><b>10.9.1</b> Best single-population task coverage</a></li>
<li class="chapter" data-level="10.9.2" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#metapopulation-task-coverage-4"><i class="fa fa-check"></i><b>10.9.2</b> Metapopulation task coverage</a></li>
</ul></li>
<li class="chapter" data-level="10.10" data-path="varied-genome-lengths.html"><a href="varied-genome-lengths.html#discussion-1"><i class="fa fa-check"></i><b>10.10</b> Discussion</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with ❤ and bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Supplemental Material for Directed Digital Evolution Project</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1 unnumbered hasAnchor">
<h1>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h1>

<div id="refs" class="references">
<div>
<p>Dolson, Emily, Alexander Lalejini, and Charles Ofria. 2019. “Exploring Genetic Programming Systems with MAP-Elites.” In <em>Genetic Programming Theory and Practice XVI</em>, edited by Wolfgang Banzhaf, Lee Spector, and Leigh Sheneman, 1–16. Cham: Springer International Publishing. <a href="https://doi.org/10.1007/978-3-030-04735-1_1">https://doi.org/10.1007/978-3-030-04735-1_1</a>.</p>
</div>
<div>
<p>Henry, Lionel, and Hadley Wickham. 2020. <em>Purrr: Functional Programming Tools</em>. <a href="https://CRAN.R-project.org/package=purrr">https://CRAN.R-project.org/package=purrr</a>.</p>
</div>
<div>
<p>Hernandez, Jose Guadalupe, Alexander Lalejini, and Emily Dolson. 2022. “What Can Phylogenetic Metrics Tell Us About Useful Diversity in Evolutionary Algorithms?” In <em>Genetic Programming Theory and Practice Xviii</em>, edited by Wolfgang Banzhaf, Leonardo Trujillo, Stephan Winkler, and Bill Worzel, 63–82. Singapore: Springer Singapore. <a href="https://doi.org/10.1007/978-981-16-8113-4_4">https://doi.org/10.1007/978-981-16-8113-4_4</a>.</p>
</div>
<div>
<p>Lalejini, Alexander M. 2022. “Selection Schemes from Evolutionary Computing Show Promise for Directed Evolution of Microbes.” OSF. <a href="https://doi.org/10.17605/OSF.IO/ZN63X">https://doi.org/10.17605/OSF.IO/ZN63X</a>.</p>
</div>
<div>
<p>Müller, Kirill, and Hadley Wickham. 2022. <em>Tibble: Simple Data Frames</em>. <a href="https://CRAN.R-project.org/package=tibble">https://CRAN.R-project.org/package=tibble</a>.</p>
</div>
<div>
<p>Neuwirth, Erich. 2022. <em>RColorBrewer: ColorBrewer Palettes</em>. <a href="https://CRAN.R-project.org/package=RColorBrewer">https://CRAN.R-project.org/package=RColorBrewer</a>.</p>
</div>
<div>
<p>R Core Team. 2022. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.</p>
</div>
<div>
<p>Wickham, Hadley. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>. Springer-Verlag New York. <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.</p>
</div>
<div>
<p>———. 2019. <em>Stringr: Simple, Consistent Wrappers for Common String Operations</em>. <a href="https://CRAN.R-project.org/package=stringr">https://CRAN.R-project.org/package=stringr</a>.</p>
</div>
<div>
<p>———. 2021a. <em>Forcats: Tools for Working with Categorical Variables (Factors)</em>. <a href="https://CRAN.R-project.org/package=forcats">https://CRAN.R-project.org/package=forcats</a>.</p>
</div>
<div>
<p>———. 2021b. <em>Tidyverse: Easily Install and Load the Tidyverse</em>. <a href="https://CRAN.R-project.org/package=tidyverse">https://CRAN.R-project.org/package=tidyverse</a>.</p>
</div>
<div>
<p>Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the tidyverse.” <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.</p>
</div>
<div>
<p>Wickham, Hadley, Winston Chang, Lionel Henry, Thomas Lin Pedersen, Kohske Takahashi, Claus Wilke, Kara Woo, Hiroaki Yutani, and Dewey Dunnington. 2022. <em>Ggplot2: Create Elegant Data Visualisations Using the Grammar of Graphics</em>. <a href="https://CRAN.R-project.org/package=ggplot2">https://CRAN.R-project.org/package=ggplot2</a>.</p>
</div>
<div>
<p>Wickham, Hadley, Romain Fran?ois, Lionel Henry, and Kirill M?ller. 2022. <em>Dplyr: A Grammar of Data Manipulation</em>. <a href="https://CRAN.R-project.org/package=dplyr">https://CRAN.R-project.org/package=dplyr</a>.</p>
</div>
<div>
<p>Wickham, Hadley, and Maximilian Girlich. 2022. <em>Tidyr: Tidy Messy Data</em>. <a href="https://CRAN.R-project.org/package=tidyr">https://CRAN.R-project.org/package=tidyr</a>.</p>
</div>
<div>
<p>Wickham, Hadley, Jim Hester, and Jennifer Bryan. 2022. <em>Readr: Read Rectangular Text Data</em>. <a href="https://CRAN.R-project.org/package=readr">https://CRAN.R-project.org/package=readr</a>.</p>
</div>
<div>
<p>Wickham, Hadley, and Dana Seidel. 2022. <em>Scales: Scale Functions for Visualization</em>. <a href="https://CRAN.R-project.org/package=scales">https://CRAN.R-project.org/package=scales</a>.</p>
</div>
<div>
<p>Wilke, Claus O. 2020. <em>Cowplot: Streamlined Plot Theme and Plot Annotations for Ggplot2</em>. <a href="https://wilkelab.org/cowplot/">https://wilkelab.org/cowplot/</a>.</p>
</div>
<div>
<p>Xie, Yihui. 2016. <em>Bookdown: Authoring Books and Technical Documents with R Markdown</em>. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://bookdown.org/yihui/bookdown">https://bookdown.org/yihui/bookdown</a>.</p>
</div>
<div>
<p>———. 2022. <em>Bookdown: Authoring Books and Technical Documents with R Markdown</em>. <a href="https://CRAN.R-project.org/package=bookdown">https://CRAN.R-project.org/package=bookdown</a>.</p>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="varied-genome-lengths.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/amlalejini/directed-digital-evolution/tree/main/tail.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
